// neo4j/lib/Node.js
/*** Generated by streamline 0.4.0 (callbacks) - DO NOT EDIT ***/
var __rt=require('streamline/lib/callbacks/runtime').runtime(__filename),__func=__rt.__func,__cb=__rt.__cb,__tryCatch=__rt.__tryCatch,__apply=__rt.__apply;
/*     1 */ (function() {
/*     2 */   var Node, Path, PropertyContainer, Relationship, adjustError, status, util, __hasProp = {
/*     3 */   }.hasOwnProperty, __extends = function(child, parent) {
/*     4 */     for (var key in parent) {
/*     4 */       if (__hasProp.call(parent, key)) {
/*     4 */         child[key] = parent[key];
                  };
                };
/*     4 */     function ctor() {
/*     4 */       this.constructor = child;
                };
/*     4 */     ctor.prototype = parent.prototype;
/*     4 */     child.prototype = new ctor();
/*     4 */     child.__super__ = parent.prototype;
/*     4 */     return child;
              };
/*     6 */   status = require("http-status");
/*     8 */   util = require("./util");
/*    10 */   adjustError = util.adjustError;
/*    12 */   PropertyContainer = require("./PropertyContainer");
/*    14 */   Relationship = require("./Relationship");
/*    16 */   Path = require("./Path");
/*    18 */   module.exports = Node = (function(_super) {
/*    20 */     __extends(Node, _super);
/*    22 */     function Node(db, data) {
/*    23 */       Node.__super__.constructor.call(this, db, data);
                };
/*    26 */     Node.prototype.toString = function() {
/*    27 */       return ("node @" + this.id);
                };
/*    30 */     Node.prototype.save = function Node_prototype_save__1(_) {
                  var message, response, services, _ref, _ref1, __this = this;
                  var __frame = {
                    name: "Node_prototype_save__1",
                    line: 30
                  };
                  return __func(_, this, arguments, Node_prototype_save__1, 0, __frame, function __$Node_prototype_save__1() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_save__1() {
                          return (function __$Node_prototype_save__1(__then) {
/*    33 */                 if (__this.exists) {
/*    34 */                   return __this._request.put({
/*    35 */                     uri: (("" + __this.self) + "/properties"),
/*    36 */                     json: __this.data
                              }, __cb(_, __frame, 4, 21, function ___(__0, __1) {
/*    34 */                     response = __1;
/*    38 */                     if ((response.statusCode !== status.NO_CONTENT)) {
/*    39 */                       message = (((_ref = response.body) != null) ? _ref.message : void 0);
/*    40 */                       switch (response.statusCode) {
/*    41 */                       case status.BAD_REQUEST:
/*    41 */                         (message || (message = "Invalid data sent"));
/*    42 */                         break;
/*    43 */                       case status.NOT_FOUND:
/*    44 */                         (message || (message = "Node not found"));
/*    45 */                       };
/*    47 */                       return _(new Error(message));
                                }
                              ;
                                __then();
                              }, true));
                            }
                             else {
/*    50 */                   return __this.db.getServices(__cb(_, __frame, 20, 21, function ___(__0, __2) {
/*    50 */                     services = __2;
/*    51 */                     return __this._request.post({
/*    52 */                       uri: services.node,
/*    53 */                       json: __this.data
                                }, __cb(_, __frame, 21, 21, function ___(__0, __3) {
/*    51 */                       response = __3;
/*    55 */                       if ((response.statusCode !== status.CREATED)) {
/*    56 */                         message = (((((_ref1 = response.body) != null) ? _ref1.message : void 0)) || "Invalid data sent");
/*    57 */                         return _(new Error(message));
                                  }
                                ;
/*    59 */                       __this._data = response.body;
                                  __then();
                                }, true));
                              }, true));
                            }
                          ;
                          })(__then);
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_save__1() {
                          if (error) {
/*    62 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*    66 */     Node.prototype["delete"] = function Node_prototype_delete__2(_, force) {
                  var relationship, relationships, _i, _len, __this = this, __arguments = arguments;
                  var __frame = {
                    name: "Node_prototype_delete__2",
                    line: 66
                  };
                  return __func(_, this, arguments, Node_prototype_delete__2, 0, __frame, function __$Node_prototype_delete__2() {
/*    68 */         if ((force == null)) {
/*    69 */           force = false;
                    }
                  ;
/*    71 */         if (!__this.exists) {
                      return _(null);
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_delete__2() {
/*    75 */               return __this.all(null, __cb(_, __frame, 9, 24, function ___(__0, __1) {
/*    75 */                 relationships = __1;
/*    76 */                 if ((relationships.length && !force)) {
/*    77 */                   return _(new Error((("Could not delete " + __this) + "; still has relationships.")));
                            }
                          ;
/*    79 */                 _i = 0;
/*    79 */                 _len = relationships.length;
                            var __5 = false;
                            return (function ___(__break) {
                              var __more;
                              var __loop = __cb(_, __frame, 0, 0, function __$Node_prototype_delete__2() {
                                __more = false;
                                if (__5) {
/*    79 */                       _i++;
                                }
                                 else {
                                  __5 = true;
                                }
                              ;
/*    79 */                     var __4 = (_i < _len);
                                if (__4) {
/*    80 */                       relationship = relationships[_i];
/*    81 */                       return relationship["delete"](__cb(_, __frame, 15, 10, function __$Node_prototype_delete__2() {
                                    while (__more) {
                                      __loop();
                                    };
                                    __more = true;
                                  }, true));
                                }
                                 else {
                                  __break();
                                }
                              ;
                              });
                              do {
                                __loop();
                              } while (__more);
                              __more = true;
                            })(__then);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_delete__2() {
                          if (error) {
/*    84 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, function __$Node_prototype_delete__2() {
/*    86 */             return __apply(__cb(_, __frame, 20, 13, _, true), Node.__super__["delete"], __this, __arguments, 0);
                      });
                    });
                  });
                };
/*    89 */     Node.prototype.del = Node.prototype["delete"];
/*    91 */     Node.prototype.createRelationshipTo = function Node_prototype_createRelationshipTo__3(otherNode, type, data, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_createRelationshipTo__3",
                    line: 91
                  };
                  return __func(_, this, arguments, Node_prototype_createRelationshipTo__3, 3, __frame, function __$Node_prototype_createRelationshipTo__3() {
/*    92 */         return __this._createRelationship(__this, otherNode, type, data, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*    95 */     Node.prototype.createRelationshipFrom = function Node_prototype_createRelationshipFrom__4(otherNode, type, data, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_createRelationshipFrom__4",
                    line: 95
                  };
                  return __func(_, this, arguments, Node_prototype_createRelationshipFrom__4, 3, __frame, function __$Node_prototype_createRelationshipFrom__4() {
/*    96 */         return __this._createRelationship(otherNode, __this, type, data, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*    99 */     Node.prototype._createRelationship = function Node_prototype__createRelationship__5(from, to, type, data, _) {
                  var createRelationshipURL, message, otherNodeURL, relationship, response, _ref, _ref1, __this = this;
                  var __frame = {
                    name: "Node_prototype__createRelationship__5",
                    line: 99
                  };
                  return __func(_, this, arguments, Node_prototype__createRelationship__5, 4, __frame, function __$Node_prototype__createRelationship__5() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype__createRelationship__5() {
/*   102 */               createRelationshipURL = from._data["create_relationship"];
/*   103 */               otherNodeURL = to.self;
                          return (function __$Node_prototype__createRelationship__5(__then) {
/*   104 */                 if ((((createRelationshipURL != null)) && otherNodeURL)) {
/*   105 */                   return __this._request.post({
/*   106 */                     url: createRelationshipURL,
/*   107 */                     json: {
/*   108 */                       to: otherNodeURL,
/*   109 */                       data: data,
/*   110 */                       type: type
                                }
                              }, __cb(_, __frame, 6, 21, function ___(__0, __1) {
/*   105 */                     response = __1;
/*   113 */                     if ((response.statusCode !== status.CREATED)) {
/*   114 */                       message = "";
/*   115 */                       switch (response.statusCode) {
/*   116 */                       case status.BAD_REQUEST:
/*   116 */                         message = ((((((_ref = response.body) != null) ? _ref.message : void 0)) || ((((_ref1 = response.body) != null) ? _ref1.exception : void 0))) || (((((((("Invalid createRelationship: " + from.id) + " ") + type) + " ") + to.id) + " w/ data: ") + (JSON.stringify(data)))));
/*   117 */                         break;
/*   118 */                       case status.CONFLICT:
/*   119 */                         message = "\"to\" node, or the node specified by the URI not found";
/*   120 */                       };
/*   122 */                       return _(new Error(message));
                                }
                              ;
/*   124 */                     relationship = new Relationship(__this.db, response.body, from, to);
/*   125 */                     return _(null, relationship);
                              }, true));
                            }
                             else {
/*   127 */                   return _(new Error("Failed to create relationship"));
                            }
                          ;
                          })(__then);
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype__createRelationship__5() {
                          if (error) {
/*   130 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   134 */     Node.prototype._getRelationships = function Node_prototype__getRelationships__6(direction, type, _) {
                  var data, prefix, relationships, relationshipsURL, resp, types, _this, __this = this;
                  var __frame = {
                    name: "Node_prototype__getRelationships__6",
                    line: 134
                  };
                  return __func(_, this, arguments, Node_prototype__getRelationships__6, 2, __frame, function __$Node_prototype__getRelationships__6() {
                    _this = __this;
/*   137 */         types = null;
/*   138 */         if ((type != null)) {
/*   139 */           types = ((type instanceof Array) ? type : [type,]);
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype__getRelationships__6() {
/*   142 */               if ((types != null)) {
/*   143 */                 prefix = __this._data[(("" + direction) + "_typed_relationships")];
/*   144 */                 relationshipsURL = ((prefix != null) ? prefix.replace("{-list|&|types}", types.join("&")) : void 0);
                          }
/*   145 */                else {
/*   146 */                 relationshipsURL = __this._data[(("" + direction) + "_relationships")];
                          }
                        ;
/*   148 */               if (!relationshipsURL) {
/*   149 */                 return _(new Error("Couldn't find URL of relationships endpoint."));
                          }
                        ;
/*   151 */               return __this._request.get(relationshipsURL, __cb(_, __frame, 17, 15, function ___(__0, __1) {
/*   151 */                 resp = __1;
/*   152 */                 if ((resp.statusCode === status.NOT_FOUND)) {
/*   153 */                   return _(new Error("Node not found."));
                            }
                          ;
/*   155 */                 if ((resp.statusCode !== status.OK)) {
/*   156 */                   return _(new Error(("Unrecognized response code: " + resp.statusCode)));
                            }
                          ;
/*   158 */                 data = JSON.parse(resp.body);
/*   159 */                 relationships = data.map(function(data) {
/*   160 */                   if ((_this.self === data.start)) {
/*   161 */                     return new Relationship(_this.db, data, _this, null);
                              }
/*   162 */                    else {
/*   163 */                     return new Relationship(_this.db, data, null, _this);
                              }
                            ;
                            });
/*   166 */                 return _(null, relationships);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype__getRelationships__6() {
                          if (error) {
/*   168 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   172 */     Node.prototype.getRelationships = function Node_prototype_getRelationships__7(type, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_getRelationships__7",
                    line: 172
                  };
                  return __func(_, this, arguments, Node_prototype_getRelationships__7, 1, __frame, function __$Node_prototype_getRelationships__7() {
/*   173 */         return __this.all(type, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   176 */     Node.prototype.outgoing = function Node_prototype_outgoing__8(type, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_outgoing__8",
                    line: 176
                  };
                  return __func(_, this, arguments, Node_prototype_outgoing__8, 1, __frame, function __$Node_prototype_outgoing__8() {
/*   177 */         return __this._getRelationships("outgoing", type, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   180 */     Node.prototype.incoming = function Node_prototype_incoming__9(type, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_incoming__9",
                    line: 180
                  };
                  return __func(_, this, arguments, Node_prototype_incoming__9, 1, __frame, function __$Node_prototype_incoming__9() {
/*   181 */         return __this._getRelationships("incoming", type, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   184 */     Node.prototype.all = function Node_prototype_all__10(type, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_all__10",
                    line: 184
                  };
                  return __func(_, this, arguments, Node_prototype_all__10, 1, __frame, function __$Node_prototype_all__10() {
/*   185 */         return __this._getRelationships("all", type, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   188 */     Node.prototype.path = function Node_prototype_path__11(to, type, direction, maxDepth, algorithm, _) {
                  var data, end, length, nodes, path, pathURL, relationships, res, start, _this, __this = this;
                  var __frame = {
                    name: "Node_prototype_path__11",
                    line: 188
                  };
                  return __func(_, this, arguments, Node_prototype_path__11, 5, __frame, function __$Node_prototype_path__11() {
                    _this = __this;
/*   191 */         if ((maxDepth == null)) {
/*   192 */           maxDepth = 1;
                    }
                  ;
/*   194 */         if ((algorithm == null)) {
/*   195 */           algorithm = "shortestPath";
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_path__11() {
/*   198 */               pathURL = (("" + __this.self) + "/path");
/*   199 */               data = {
/*   200 */                 to: to.self,
/*   201 */                 relationships: {
/*   202 */                   type: type,
/*   203 */                   direction: direction
                            },
/*   205 */                 max_depth: maxDepth,
/*   206 */                 algorithm: algorithm
                          };
/*   208 */               return __this._request.post({
/*   209 */                 url: pathURL,
/*   210 */                 json: data
                          }, __cb(_, __frame, 20, 14, function ___(__0, __1) {
/*   208 */                 res = __1;
/*   212 */                 if ((res.statusCode === status.NOT_FOUND)) {
/*   213 */                   return _(null, null);
                            }
                          ;
/*   215 */                 if ((res.statusCode !== status.OK)) {
/*   216 */                   return _(new Error(("Unrecognized response code: " + res.statusCode)));
                            }
                          ;
/*   218 */                 data = res.body;
/*   219 */                 start = new Node(__this, {
/*   220 */                   self: data.start
                            });
/*   222 */                 end = new Node(__this, {
/*   223 */                   self: data.end
                            });
/*   225 */                 length = data.length;
/*   226 */                 nodes = data.nodes.map(function(url) {
/*   227 */                   return new Node(_this, {
/*   228 */                     self: url
                              });
                            });
/*   231 */                 relationships = data.relationships.map(function(url) {
/*   232 */                   return new Relationship(_this, {
/*   233 */                     self: url,
/*   234 */                     type: type
                              });
                            });
/*   237 */                 path = new Path(start, end, length, nodes, relationships);
/*   238 */                 return _(null, path);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_path__11() {
                          if (error) {
/*   240 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   244 */     Node.prototype.getRelationshipNodes = function Node_prototype_getRelationshipNodes__12(rels, _) {
                  var resp, traverseURL, _ref, _ref1, _ref2, _this, __this = this;
                  var __frame = {
                    name: "Node_prototype_getRelationshipNodes__12",
                    line: 244
                  };
                  return __func(_, this, arguments, Node_prototype_getRelationshipNodes__12, 1, __frame, function __$Node_prototype_getRelationshipNodes__12() {
                    _this = __this;
/*   247 */         rels = ((rels instanceof Array) ? rels : [rels,]);
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_getRelationshipNodes__12() {
/*   249 */               traverseURL = (((_ref = __this._data["traverse"]) != null) ? _ref.replace("{returnType}", "node") : void 0);
/*   250 */               if (!traverseURL) {
/*   251 */                 return _(new Error("Traverse not available."));
                          }
                        ;
/*   253 */               return __this._request.post({
/*   254 */                 url: traverseURL,
/*   255 */                 json: {
/*   256 */                   max_depth: 1,
/*   257 */                   relationships: rels.map(function(rel) {
/*   258 */                     if ((typeof rel === "string")) {
/*   259 */                       return {
/*   260 */                         type: rel
                                  };
                                }
/*   262 */                      else {
/*   263 */                       return rel;
                                }
                              ;
                              })
                            }
                          }, __cb(_, __frame, 9, 15, function ___(__0, __1) {
/*   253 */                 resp = __1;
/*   268 */                 if ((resp.statusCode === 404)) {
/*   269 */                   return _(new Error((((((_ref1 = resp.body) != null) ? _ref1.message : void 0)) || "Node not found.")));
                            }
                          ;
/*   271 */                 if ((resp.statusCode !== 200)) {
/*   272 */                   return _(new Error((((((_ref2 = resp.body) != null) ? _ref2.message : void 0)) || (("Unrecognized response code: " + resp.statusCode)))));
                            }
                          ;
/*   274 */                 return _(null, resp.body.map(function(data) {
/*   275 */                   return new Node(_this.db, data);
                            }));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_getRelationshipNodes__12() {
                          if (error) {
/*   278 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   282 */     Node.prototype.index = function Node_prototype_index__13(index, key, value, _) {
                  var encodedKey, encodedValue, response, services, url, version, __this = this;
                  var __frame = {
                    name: "Node_prototype_index__13",
                    line: 282
                  };
                  return __func(_, this, arguments, Node_prototype_index__13, 3, __frame, function __$Node_prototype_index__13() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_index__13() {
/*   285 */               if (!__this.exists) {
/*   286 */                 return _(new Error("Node must exists before indexing properties"));
                          }
                        ;
/*   288 */               return __this.db.getServices(__cb(_, __frame, 6, 19, function ___(__0, __1) {
/*   288 */                 services = __1;
/*   289 */                 return __this.db.getVersion(__cb(_, __frame, 7, 18, function ___(__0, __2) {
/*   289 */                   version = __2;
                              return (function __$Node_prototype_index__13(__then) {
/*   290 */                     if ((version <= 1.4)) {
/*   291 */                       encodedKey = encodeURIComponent(key);
/*   292 */                       encodedValue = encodeURIComponent(value);
/*   293 */                       url = ((((((("" + services.node_index) + "/") + index) + "/") + encodedKey) + "/") + encodedValue);
/*   294 */                       return __this._request.post({
/*   295 */                         url: url,
/*   296 */                         json: __this.self
                                  }, __cb(_, __frame, 12, 21, function ___(__0, __3) {
/*   294 */                         response = __3;
                                    __then();
                                  }, true));
                                }
                                 else {
/*   299 */                       return __this._request.post({
/*   300 */                         url: ((("" + services.node_index) + "/") + index),
/*   301 */                         json: {
/*   302 */                           key: key,
/*   303 */                           value: value,
/*   304 */                           uri: __this.self
                                    }
                                  }, __cb(_, __frame, 17, 21, function ___(__0, __4) {
/*   299 */                         response = __4;
                                    __then();
                                  }, true));
                                }
                              ;
                              })(function __$Node_prototype_index__13() {
/*   308 */                     if ((response.statusCode !== status.CREATED)) {
/*   309 */                       return _(response);
                                }
                              ;
                                __then();
                              });
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_index__13() {
                          if (error) {
/*   312 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   316 */     return Node;
/*   318 */   })(PropertyContainer);
/*   320 */ }).call(this);
