// neo4j/lib/GraphDatabase.js
/*** Generated by streamline 0.4.0 (callbacks) - DO NOT EDIT ***/
var __rt=require('streamline/lib/callbacks/runtime').runtime(__filename),__func=__rt.__func,__cb=__rt.__cb,__tryCatch=__rt.__tryCatch;
/*     1 */ (function() {
/*     2 */   var GraphDatabase, Node, Relationship, adjustError, status, util;
/*     4 */   status = require("http-status");
/*     6 */   util = require("./util");
/*     8 */   adjustError = util.adjustError;
/*    10 */   Relationship = require("./Relationship");
/*    12 */   Node = require("./Node");
/*    14 */   module.exports = GraphDatabase = (function() {
/*    15 */     var _this = this;
/*    17 */     function GraphDatabase(url) {
/*    18 */       this.url = url;
/*    19 */       this._request = util.wrapRequestForAuth(url);
/*    20 */       this._root = null;
/*    21 */       this._services = null;
                };
/*    24 */     GraphDatabase.prototype._purgeCache = function() {
/*    25 */       this._root = null;
/*    26 */       return this._services = null;
                };
/*    29 */     GraphDatabase.prototype._getRoot = function GraphDatabase_prototype__getRoot__1(_) {
                  var response, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype__getRoot__1",
                    line: 29
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype__getRoot__1, 0, __frame, function __$GraphDatabase_prototype__getRoot__1() {
/*    31 */         if ((__this._root != null)) {
/*    32 */           return _(null, __this._root);
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype__getRoot__1() {
/*    35 */               return __this._request.get(__this.url, __cb(_, __frame, 6, 19, function ___(__0, __1) {
/*    35 */                 response = __1;
/*    36 */                 if ((response.statusCode !== status.OK)) {
/*    37 */                   return _(response);
                            }
                          ;
/*    39 */                 __this._root = JSON.parse(response.body);
/*    40 */                 return _(null, __this._root);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype__getRoot__1() {
                          if (error) {
/*    42 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*    46 */     GraphDatabase.prototype.getServices = function GraphDatabase_prototype_getServices__2(_) {
                  var response, root, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getServices__2",
                    line: 46
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getServices__2, 0, __frame, function __$GraphDatabase_prototype_getServices__2() {
/*    48 */         if ((__this._services != null)) {
/*    49 */           return _(null, __this._services);
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getServices__2() {
/*    52 */               return __this._getRoot(__cb(_, __frame, 6, 15, function ___(__0, __1) {
/*    52 */                 root = __1;
/*    53 */                 return __this._request.get(root.data, __cb(_, __frame, 7, 19, function ___(__0, __2) {
/*    53 */                   response = __2;
/*    54 */                   if ((response.statusCode !== status.OK)) {
/*    55 */                     return _(response);
                              }
                            ;
/*    57 */                   __this._services = JSON.parse(response.body);
/*    58 */                   return _(null, __this._services);
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getServices__2() {
                          if (error) {
/*    60 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*    64 */     GraphDatabase.prototype.getVersion = function GraphDatabase_prototype_getVersion__3(_) {
                  var services, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getVersion__3",
                    line: 64
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getVersion__3, 0, __frame, function __$GraphDatabase_prototype_getVersion__3() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getVersion__3() {
/*    67 */               return __this.getServices(__cb(_, __frame, 3, 19, function ___(__0, __1) {
/*    67 */                 services = __1;
/*    68 */                 return _(null, parseFloat((services["neo4j_version"] || "1.4")));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getVersion__3() {
                          if (error) {
/*    70 */                 return _(adjustError);
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*    74 */     GraphDatabase.prototype.createNode = function(data) {
/*    75 */       var node;
/*    76 */       data = (data || {
                  });
/*    77 */       node = new Node(this, {
/*    78 */         data: data
                  });
/*    80 */       return node;
                };
/*    83 */     GraphDatabase.prototype.getNode = function GraphDatabase_prototype_getNode__4(url, _) {
                  var node, response, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getNode__4",
                    line: 83
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getNode__4, 1, __frame, function __$GraphDatabase_prototype_getNode__4() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getNode__4() {
/*    86 */               return __this._request.get(url, __cb(_, __frame, 3, 19, function ___(__0, __1) {
/*    86 */                 response = __1;
/*    87 */                 if ((response.statusCode !== status.OK)) {
/*    88 */                   if ((response.statusCode === status.NOT_FOUND)) {
/*    89 */                     return _(new Error(("No node at " + url)));
                              }
                            ;
/*    91 */                   return _(response);
                            }
                          ;
/*    93 */                 node = new Node(__this, JSON.parse(response.body));
/*    94 */                 return _(null, node);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getNode__4() {
                          if (error) {
/*    96 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   100 */     GraphDatabase.prototype.getIndexedNode = function GraphDatabase_prototype_getIndexedNode__5(index, property, value, _) {
                  var node, nodes, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getIndexedNode__5",
                    line: 100
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getIndexedNode__5, 3, __frame, function __$GraphDatabase_prototype_getIndexedNode__5() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getIndexedNode__5() {
/*   103 */               return __this.getIndexedNodes(index, property, value, __cb(_, __frame, 3, 16, function ___(__0, __1) {
/*   103 */                 nodes = __1;
/*   104 */                 node = null;
/*   105 */                 if ((nodes && (nodes.length > 0))) {
/*   106 */                   node = nodes[0];
                            }
                          ;
/*   108 */                 return _(null, node);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getIndexedNode__5() {
                          if (error) {
/*   110 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   114 */     GraphDatabase.prototype.getIndexedNodes = function GraphDatabase_prototype_getIndexedNodes__6(index, property, value, _) {
                  var key, nodeArray, nodes, response, services, url, val, _this, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getIndexedNodes__6",
                    line: 114
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getIndexedNodes__6, 3, __frame, function __$GraphDatabase_prototype_getIndexedNodes__6() {
                    _this = __this;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getIndexedNodes__6() {
/*   118 */               return __this.getServices(__cb(_, __frame, 4, 19, function ___(__0, __1) {
/*   118 */                 services = __1;
/*   119 */                 key = encodeURIComponent(property);
/*   120 */                 val = encodeURIComponent(value);
/*   121 */                 url = ((((((("" + services.node_index) + "/") + index) + "/") + key) + "/") + val);
/*   122 */                 return __this._request.get(url, __cb(_, __frame, 8, 19, function ___(__0, __2) {
/*   122 */                   response = __2;
/*   123 */                   if ((response.statusCode !== status.OK)) {
/*   124 */                     return _(response);
                              }
                            ;
/*   126 */                   nodeArray = JSON.parse(response.body);
/*   127 */                   nodes = nodeArray.map(function(node) {
/*   128 */                     return new Node(_this, node);
                              });
/*   130 */                   return _(null, nodes);
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getIndexedNodes__6() {
                          if (error) {
/*   132 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   136 */     GraphDatabase.prototype.getNodeById = function GraphDatabase_prototype_getNodeById__7(id, _) {
                  var node, services, url, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getNodeById__7",
                    line: 136
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getNodeById__7, 1, __frame, function __$GraphDatabase_prototype_getNodeById__7() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getNodeById__7() {
/*   139 */               return __this.getServices(__cb(_, __frame, 3, 19, function ___(__0, __1) {
/*   139 */                 services = __1;
/*   140 */                 url = ((("" + services.node) + "/") + id);
/*   141 */                 return __this.getNode(url, __cb(_, __frame, 5, 15, function ___(__0, __2) {
/*   141 */                   node = __2;
/*   142 */                   return _(null, node);
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getNodeById__7() {
                          if (error) {
/*   144 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   148 */     GraphDatabase.prototype.createRelationship = function GraphDatabase_prototype_createRelationship__8(startNode, endNode, type, _) {
                  var __frame = {
                    name: "GraphDatabase_prototype_createRelationship__8",
                    line: 148
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_createRelationship__8, 3, __frame, _);
                };
/*   150 */     GraphDatabase.prototype.getRelationship = function GraphDatabase_prototype_getRelationship__9(url, _) {
                  var data, relationship, response, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getRelationship__9",
                    line: 150
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getRelationship__9, 1, __frame, function __$GraphDatabase_prototype_getRelationship__9() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getRelationship__9() {
/*   153 */               return __this._request.get(url, __cb(_, __frame, 3, 19, function ___(__0, __1) {
/*   153 */                 response = __1;
/*   154 */                 if ((response.statusCode !== status.OK)) {
/*   155 */                   return _(response);
                            }
                          ;
/*   157 */                 data = JSON.parse(response.body);
/*   158 */                 relationship = new Relationship(__this, data);
/*   159 */                 return _(null, relationship);
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_getRelationship__9() {
                          if (error) {
/*   161 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   165 */     GraphDatabase.prototype.getRelationshipById = function GraphDatabase_prototype_getRelationshipById__10(id, _) {
                  var relationshipURL, services, url, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_getRelationshipById__10",
                    line: 165
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_getRelationshipById__10, 1, __frame, function __$GraphDatabase_prototype_getRelationshipById__10() {
/*   167 */         return __this.getServices(__cb(_, __frame, 2, 17, function ___(__0, __1) {
/*   167 */           services = __1;
/*   168 */           relationshipURL = services.node.replace("node", "relationship");
/*   169 */           url = ((("" + relationshipURL) + "/") + id);
/*   170 */           return __this.getRelationship(url, __cb(_, __frame, 5, 13, _, true));
                    }, true));
                  });
                };
/*   173 */     GraphDatabase.prototype.query = function GraphDatabase_prototype_query__11(query, params, _) {
                  var body, columns, endpoint, i, map, response, results, row, services, value, _ref, _ref1, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_query__11",
                    line: 173
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_query__11, 2, __frame, function __$GraphDatabase_prototype_query__11() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_query__11() {
/*   176 */               return __this.getServices(__cb(_, __frame, 3, 19, function ___(__0, __2) {
/*   176 */                 services = __2;
/*   177 */                 endpoint = (services.cypher || ((((_ref = services.extensions) != null) ? (((_ref1 = _ref.CypherPlugin) != null) ? _ref1["execute_query"] : void 0) : void 0)));
/*   178 */                 if (!endpoint) {
/*   179 */                   return _(new Error("Cypher plugin not installed"));
                            }
                          ;
/*   181 */                 return __this._request.post({
/*   182 */                   uri: endpoint,
/*   183 */                   json: (params ? {
/*   184 */                     query: query,
/*   185 */                     params: params
/*   186 */                   } : {
/*   187 */                     query: query
                              })
                            }, __cb(_, __frame, 8, 19, function ___(__0, __3) {
/*   181 */                   response = __3;
/*   190 */                   if ((response.statusCode === status.NO_CONTENT)) {
/*   191 */                     return _(new Error((("Unknown Neo4j error for query:\n\n" + query) + "\n")));
                              }
                            ;
/*   193 */                   if ((response.statusCode !== status.OK)) {
/*   194 */                     return _(response);
                              }
                            ;
/*   196 */                   body = response.body;
/*   197 */                   columns = body.columns;
/*   212 */                   return (function __1(_) {
                                var _i, _j, _len, _len1, _ref2, _results;
/*   200 */                     _ref2 = body.data;
/*   201 */                     _results = [];
/*   202 */                     for (_i = 0, _len = _ref2.length; (_i < _len); _i++) {
/*   203 */                       row = _ref2[_i];
/*   204 */                       map = {
                                  };
/*   205 */                       for (i = _j = 0, _len1 = row.length; (_j < _len1); i = ++_j) {
/*   206 */                         value = row[i];
/*   207 */                         map[columns[i]] = (((value && (typeof value === "object")) && value.self) ? (value.type ? new Relationship(__this, value) : new Node(__this, value)) : value);
                                  };
/*   209 */                       _results.push(map);
                                };
/*   211 */                     return _(null, _results);
/*   212 */                   })(__cb(_, __frame, 39, 18, function ___(__0, __4) {
/*   198 */                     results = __4;
/*   213 */                     return _(null, results);
                              }, true));
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_query__11() {
                          if (error) {
/*   215 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   219 */     (function(actual) {
/*   220 */       return GraphDatabase.prototype.query = function(query, params, callback) {
/*   221 */         if (((typeof query === "function") && (typeof params === "string"))) {
/*   222 */           console.warn((("neo4j.GraphDatabase::query()’s signature is " + "now (query, params, callback). Please update your code!\n") + new Error().stack.split("\n")[2]));
/*   223 */           callback = query;
/*   224 */           query = params;
/*   225 */           params = null;
                    }
/*   226 */          else if ((typeof params === "function")) {
/*   227 */           callback = params;
/*   228 */           params = null;
                    }
                    
                  ;
/*   230 */         return actual.call(this, query, params, callback);
                  };
/*   232 */     })(GraphDatabase.prototype.query);
/*   234 */     GraphDatabase.prototype.queryNodeIndex = function GraphDatabase_prototype_queryNodeIndex__12(index, query, _) {
                  var nodeArray, nodes, response, services, url, _this, __this = this;
                  var __frame = {
                    name: "GraphDatabase_prototype_queryNodeIndex__12",
                    line: 234
                  };
                  return __func(_, this, arguments, GraphDatabase_prototype_queryNodeIndex__12, 2, __frame, function __$GraphDatabase_prototype_queryNodeIndex__12() {
                    _this = __this;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$GraphDatabase_prototype_queryNodeIndex__12() {
/*   238 */               return __this.getServices(__cb(_, __frame, 4, 19, function ___(__0, __1) {
/*   238 */                 services = __1;
/*   239 */                 url = ((((("" + services.node_index) + "/") + index) + "?query=") + (encodeURIComponent(query)));
/*   240 */                 return __this._request.get(url, __cb(_, __frame, 6, 19, function ___(__0, __2) {
/*   240 */                   response = __2;
/*   241 */                   if ((response.statusCode !== status.OK)) {
/*   242 */                     return _(response);
                              }
                            ;
/*   244 */                   nodeArray = JSON.parse(response.body);
/*   245 */                   nodes = nodeArray.map(function(node) {
/*   246 */                     return new Node(_this, node);
                              });
/*   248 */                   return _(null, nodes);
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$GraphDatabase_prototype_queryNodeIndex__12() {
                          if (error) {
/*   250 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   254 */     return GraphDatabase;
/*   256 */   }).call(this);
/*   258 */ }).call(this);
